src/aspire/noise/noise.py

```diff
@@ -345,3 +354,297 @@ def estimate_noise_psd(self):
+            s = fft.irfft2(ftmp * ftmp.conj(), s=tmp.shape[1:])
+            s = s[:, 0 : max_d + 1, 0 : max_d + 1]  # crop
+
+            # Accumulate all autocorrelation values R[k1,k2] such that
+            # k1**2 + k2**2 = dist (all autocorrelations of a certain distance).
+            s = xp.sum(s, axis=0).flatten()
+            _Ncorr = Ncorr.flatten() * count
+            for d in validdists:
+                idx = distmap[d]
+                corrs[idx] = corrs[idx] + s[d]
+                corrcount[idx] = corrcount[idx] + _Ncorr[d]
+
+        # Remove distances which had no samples
+        idx = xp.where(corrcount != 0)
+        R = corrs[idx] / corrcount[idx]
+        x = xp.asnumpy(x[idx])
+        cnt = corrcount[idx]
+
+        R = xp.asnumpy(R)
+        cnt = xp.asnumpy(cnt)
+
+        return R, x, cnt
+
+    @staticmethod
+    def estimate_power_spectrum_distribution_2d(
+        images, samples_idx, max_d=None, batch_size=512
+    ):
+        """
+        Estimate the 2D isotropic power spectrum of `images`.
+        The samples to use in each image are given by `samples_idx` mask.
+        The correlation is computed up to a maximal distance of `max_d`.
+
+        Port of MATLAB `cryo_epsdS`.
+
+        :param images: Images instance
+        :param samples_idx: Boolean mask shaped (L,L).
+        :param max_d: Max computed correlation distance in pixels.
+           Default of `None` yields `np.floor(L / 3)`.
+        :param batch_size:  The size of the batches in which to compute the variance estimate.
+        :return:
+            - 2D PSD array
+            - Radial PSD array
+            - Distances map array
+            - Count of nonzero correlations array
+        """
+
+        n_img = images.n_images
+        L = samples_idx.shape[-1]
+        batch_size = min(batch_size, n_img)
+
+        # Migrate mask to GPU as needed
+        _samples_idx = xp.asarray(samples_idx)
+
+        # Correlations more than `max_d` pixels apart are not computed.
+        if max_d is None:
+            max_d = np.floor(L / 3)
+        if max_d > L - 1:
+            logger.info(
+                f"`max_d` value {max_d}greater than number of image pixels {L}, clipping to {L-1}."
+            )
+        max_d = int(min(max_d, L - 1))
+
+        R, x, _ = LegacyNoiseEstimator.estimate_power_spectrum_distribution_1d(
+            images=images, samples_idx=samples_idx, max_d=max_d, batch_size=batch_size
+        )
+        _R = xp.asarray(R)  # Migrate to GPU for assignments below
+
+        # Use the 1D autocorrelation estimated above to populate an
+        # array of the 2D isotropic autocorrelction. This
+        # autocorrelation is later Fourier transformed to get the
+        # power spectrum.
+        R2 = xp.zeros((2 * L - 1, 2 * L - 1), dtype=np.float64)
+
+        X, Y = xp.mgrid[-L + 1 : L, -L + 1 : L]
+        dists2 = X * X + Y * Y
+        dsquare2 = xp.sort(xp.unique(dists2[dists2 <= max_d**2]))
```

Comment:
`unique_dists2`?

Reply:
I changed them to match `_1d`.  `dsquare` and `uniq_dsquare`.

I know its duplicated.  I didn't think it was worth to carve it out...

Reply:
No that's fine. Just the name I was after.